repair: Introduce new primary replica selection algorithm for tablets

bhalevy opened the pull request:
Tablet allocation does not guarantee fairness of
the first replica in the replicas set across dcs.
The lack of this fix cause the following dtest to fail:
repair_additional_test.py::TestRepairAdditional::test_repair_option_pr_multi_dc

Use the tablet_map get_primary_replica or get_primary_replica_within_dc,
respectively to see if this node is the primary replica for each tablet
or not.

Fixes https://github.com/scylladb/scylladb/issues/17752

No backport is required before 6.0 as tablets (and tablet repair) are introduced in 6.0

bhalevy commented:
This is a reincarnation of https://github.com/scylladb/scylladb/pull/18396 that I closed and I cannot reopen it.

In v4 (30d08f3bfb):
- reimplemented using tablet_map::get_primary_replica
- locator/tablets: tmap.get_primary_replica: return whole tablet_replica
- locator/tablets: add get_primary_replica_within_dc
- repair: repair_tablets: no need to check ranges_specified per tablet
- repair: repair_tablets: select primary dc per tablet
  - now using that tablet_map get_primary_replica or get_primary_replica_within_dc

asias commented on locator/tablets.cc:
```diff
@@ -245,9 +246,19 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+tablet_replica tablet_map::get_primary_replica(tablet_id id) const {
     const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+    return info.replicas.at(size_t(id) % info.replicas.size());
```
I am wondering if this gives enough randomness. 

Consider:

tablet0   n1, n2, n3             (n1 is the primary replica)
tablet1    n2, n1, n3             (n1 is the primary replica)
tablet2   n2, n3, n1                (n1 is the primary replica)

In this case, n1 is always the primary replica.

bhalevy replied:
It's true that the node order is shuffled in the replica set so it could result in such unfairness.
We can sort it in a predetermined order, say by host_id order and then select the primary using the ordinal tablet_id modulo number of replicas, and for the within_dc case, also filter the nodes for sorting based on the given dc.

bhalevy replied:
done in v5

asias commented on repair/repair.cc:
```diff
@@ -2143,28 +2144,27 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
+                }
+                return make_ready_future<>();
+            }
+
```
We could share the meta.push_back with the non primary_replica_only case like below.  But it is up to you. 
```
        co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) -> future<> {
            auto range = tmap.get_token_range(id);
            auto& replicas = info.replicas;
            bool found = false;
            shard_id master_shard_id;

            if (primary_replica_only) {
                auto primary_replica = select_primary_ranges_within_dc ?
                        tmap.get_primary_replica_within_dc(id, erm->get_topology(), mydc) :
                        tmap.get_primary_replica(id);
                if (primary_replica.host == myhostid) {
                    found = true;
                    master_shard_id = primary_replica.shard,
                }
            } else {
                // Repair all tablets belong to this node
                for (auto& r : replicas) {
                    if (r.host == myhostid) {
                        master_shard_id = r.shard;
                        found = true;
                        break;
                    }
                }
            }
            if (found) {
                metas.push_back(repair_tablet_meta{id, range, myhostid, master_shard_id, replicas});
            }
            return make_ready_future<>();
        });
```

bhalevy replied:
right, this cross my mind too, but I doubt it's really worth it just to remain closer the the existing implementation.
We can also keep an `optional<tablet_replica>` variable instead of `bool found` and `master_shard_id`. I think that would be more elegant, but again, I'm not sure it's worth the complication in logic.

asias replied:
OK. I am good with your version too.

bhalevy commented:
In v5 ([0ac5b16](https://github.com/scylladb/scylladb/commit/0ac5b1664fdb06fe65d731056dfb92194757caf4)):
- tablet_map: get_primary_replica: pre-sort the replicas
- and similarly, do the same in get_primary_replica_within_dc

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9043/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9043/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9693/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1669/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9043/)
#### Build Details:
- Duration: 4 hr 59 min
- Builder: i-0e43e4ccfce348b4f (i4i.8xlarge)

asias commented on locator/tablets.cc:
```diff
@@ -245,9 +246,26 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+tablet_replica tablet_map::get_primary_replica(tablet_id id) const {
     const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+    auto replicas = boost::copy_range<std::vector<tablet_replica>>(info.replicas);
```
How about sorting when replicas is updated? We might use get_primary_replica in data path later.

bhalevy replied:
We can't do that, unless MV has a better way to pair replicas in the base table with the respective ones in the view.
See https://github.com/scylladb/scylladb/issues/17147, https://github.com/scylladb/scylladb/issues/17043, https://github.com/scylladb/scylladb/issues/17265

bhalevy replied:
> How about sorting when replicas is updated? We might use get_primary_replica in data path later.

Regarding the primary replica in the data path.
We don't have a crisp design for that yet, but my assumption is that when we have alignment between raft groups and tablets, e.g. for LWT over raft, we will have a raft group leader per tablet and it can be seen as the primary replica.  With hierarchical raft, you'd also have a sub-group leader per dc, and that could be seen as the primary replica within the dc.

Cc @tgrabiec

asias replied:
I see.

denesb replied:
Why the copy? And if the copy is needed, why not just:
```c++
auto replicas = std::vector(info.replicas.begin(), info.replicas.end());
```
Let's keep things simple.

bhalevy replied:
As I answered [above](https://github.com/scylladb/scylladb/pull/18784/files#r1609386314), the copy is needed since we mustn't sort the replicas in place.

The constructor you suggested makes sense and it is indeed simple.
I'll send a new version, also with the partial sort that should be more efficient on average for this purpose, and with the expected relatively low number of replicas.
replica count is number of dcs * rf_per_dc and I think we can safely assume this will be bound by 5 * 5 = 25 replicas, in very large installations

bhalevy commented:
@denesb please review too

raphaelsc commented on locator/tablets.cc:
```diff
@@ -245,9 +246,26 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+        throw std::runtime_error(format("Cannot get primary replica for tablet_id={} in dc={}: no replicas found in this datacenter", id, dc));
+    }
+    std::sort(replicas.begin(), replicas.end(), [] (const tablet_replica& l, const tablet_replica& r) {
+        return l.host < r.host;
```
auto idx = size_t(id) % replicas.size();
std::partial_sort(replicas, replicas.begin() + (idx + 1));

bhalevy replied:
interesting idea.
And we can sort in reverse if idx is greater than the size/2

asias replied:
Don't we need a stable sort?

The doc for std::partial_sort says:

The order of equal elements is not guaranteed to be preserved. The order of the remaining elements in the range [middle, last) is unspecified.

raphaelsc replied:
after the partial sort, the elements in [middle, last) are guaranteed to be ordered after [first, middle). and preexisting order of equal elements is not a requirement for us, as replicas will have different hosts. the idea is to reduce complexity from O(n * log n) to O(n * log m), where m = distance from first element to middle one.

asias replied:
Yes, I know the idea behind the partial sort.  I have concerns non stable sort would choose the wrong host. 

Assume n1 == n2 in sorting.

before sort: n4 n3 n1 n2

after stable sort:
n1, n2, n3, n4

1) after non stable sort version 1:
n2, n1, n3, n4

2) after non stable sort version 2:
n1, n2, n3, n4

for 1) with idx = 0, we choose n2
for 2) with idx = 0, we choose n1

raphaelsc replied:
I think you misinterpreted 'The order of the remaining elements in the range [middle, last) is unspecified.'. We're not interested in [middle, last), that's beyond the index to be used (size_t(id) % replicas.size()). The order of elements in the range [first, middle) is not unspecified, they will be sorted.

raphaelsc replied:
Ah ok it's me who misinterpreted you, it seems. I don't follow this 'n1 == n2' too late for me. will leave the judgment for Benny. If this is not a hot path, optimizating it doesn't matter.

asias replied:
Nope, I am not talking about the remaining elements. I am talking about the sorted ones. The sorted ones do not preserve the order for equal elements.

asias replied:
However, since we use a unique id for sorting, the equal case will not happen. Not preserving the order is not a problem for this use case.

denesb reviewed (commented):
LGTM, one nit.

bhalevy commented:
In v6 ([802704f](https://github.com/scylladb/scylladb/commit/802704f3f0969c1350fd21f8c67fbc217a482d03)):
- simplified constructor for copying the replicas vector
- used a static function to sort the replica set and select the primary replica
  - called from both get_primary_replica and get_primary_replica_within_dc
  - uses std::partial_sort for either the vector prefix or suffix, depending on the tablet idx

asias commented:
@bhalevy we need to rename the title of this PR.

bhalevy commented:
> @bhalevy we need to rename the title of this PR.

what would you suggest?

asias commented on locator/tablets.cc:
```diff
@@ -245,9 +246,42 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+        });
+    } else {
+        // sort the last (size - idx) replicas in descending order
+        std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx + 1, replicas.rend(), [] (const tablet_replica& l, const tablet_replica& r) {
```
I think this is incorrect. 

We need to sort either
[0, idx] (first part) or [idx, size-1] (last part).

For [idx, size -1 ], we make sure the last n = size -1 - idx + 1 = size - idx elements are sorted.

The code above sorts the last size - idx + 1 + 1 = size -  idx + 2 elements, which is wrong. 

It should be std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx  -1, replicas.rend() ...

bhalevy replied:
hmm, I think your calculation is off by one, to the other (unsafe) side.

partial_sort sorts `[begin, middle)`, so when sorting the vector tail, say size=6, idx=3:

|  | 0 | 1 | 2 | 3 | 4 | 5 |
|---|---|---|---|---|---|---|
| rend |   |   |  mid |   |  | rbegin |

we want mid = rbegin + 3 = rbegin + (size - idx)

If idx=5:

|  | 0 | 1 | 2 | 3 | 4 | 5 |
|---|---|---|---|---|---|---|
| rend |   |   |    |   | mid | rbegin |

we want mid = rbegin + 1 = rbegin + (size - idx)

asias replied:
Correct, since (rbegin + midlle  - rbegin  + 1  - 1 = size - idx) , so  middle  = size -idx for the reverse sort.

asias commented:
> > @bhalevy we need to rename the title of this PR.
> 
> what would you suggest?

repair: Introduce new primary replica selection algorithm for tablet

?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1702/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9726/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/)
#### Build Details:
- Duration: 4 hr 32 min
- Builder: i-0b850fa4f1da05c40 (i4i.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1702/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9726/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9120/)
> 
> #### Build Details:
> * Duration: 4 hr 32 min
> * Builder: i-0b850fa4f1da05c40 (i4i.8xlarge)

unrelated issue: https://github.com/scylladb/scylladb/issues/15877

bhalevy commented:
In v7 ([ca32e60](https://github.com/scylladb/scylladb/commit/ca32e60f9d8f21de8aa6d09fd6ca87f0f218df66)):
- fixed partial_sort mid iterator calculation as per https://github.com/scylladb/scylladb/pull/18784#discussion_r1611754744
- moved static function to be a static method and added a unit test for it

[tablet-repair-primary-range-diff-v6-v7.patch.txt](https://github.com/scylladb/scylladb/files/15418589/tablet-repair-primary-range-diff-v6-v7.patch.txt)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9130/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9130/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test::*
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9130/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1710/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9735/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9130/)
#### Build Details:
- Duration: 8 hr 48 min
- Builder: i-074e155db88ad1c49 (r5ad.8xlarge)

asias commented on locator/tablets.cc:
```diff
@@ -245,9 +246,42 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+const tablet_replica& tablet_map::sort_and_get_primary_replica(tablet_replica_set& replicas, tablet_id id) {
```
Passing tablet_replica_set& which allows modifying the replicas at the caller side looks dangerous. People might misuse it to sort the info.replicas unintentionally. 

How about:

tablet_replica tablet_map::get_primary_replica(const tablet_replica_set& replicas, tablet_id id). 

This is enough even for the unitest below.

bhalevy replied:
And let repair di the filtering for the dc?
That's possible.  Though a bit of an overkill since this is a private method by design and not intended for public use.
Also, its name suggests it's sorting the input vector so it shouldn't come as a surprise to the caller.

asias replied:
I meant we keep the exiting 

tablet_replica tablet_map::get_primary_replica_within_dc(tablet_id id, const topology& topo, const sstring& dc) tablet_replica tablet_map::get_primary_replica(tablet_id id) const

in addition, we have

tablet_replica tablet_map::get_primary_replica( tablet_id id, const tablet_replica_set& replicas)

which can be used to get tablet_replica with the given replicas provided. 

The filter stuff is still done by the  tablet_replica tablet_map::get_primary_replica_within_dc.

bhalevy replied:
This will add yet another copy of the replicas within the dc.
To save it, the low-level get_primary_replica function can receive also a predicate function to filter the `const tablet_replica_set& replica` while copying it, before sorting the copy in place.

avikivity commented on locator/tablets.cc:
```diff
@@ -245,9 +245,32 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+        std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx, replicas.rend(), [] (const tablet_replica& l, const tablet_replica& r) {
+            return l.host > r.host;
+        });
+    }
```
Are you looking for nth_element?

Note std::ranges::partial_sort accepts a projection so you can pass std::mem_fn(&tablet_replica::host).

I'm not convinced this is less vulnerable to bad distributions than the original.

bhalevy replied:
> Are you looking for nth_element?

yes

> Note std::ranges::partial_sort accepts a projection so you can pass std::mem_fn(&tablet_replica::host).


That would be std::ranges::partial_sort, but yes, it would simplify this a bit.

> I'm not convinced this is less vulnerable to bad distributions than the original.

We can consider the nodes' network topology so to rotate DCs and RACKs before nodes (in host_id, or node->idx order).

bhalevy replied:
> but I'm not sure it's worth it to lookup the node for each host, for each tablet.

Hmm, we do it for get_primary_replica_within_dc anyways...  so maybe it isn't so bad.

avikivity replied:
> 
> > I'm not convinced this is less vulnerable to bad distributions than the original.
> 
> We can consider the nodes' network topology so to rotate DCs and RACKs before nodes (in host_id, or node->idx order).

You're fixing a problem I don't understand. Yes, if the tablet replicas are exactly the wrong way, we get a lopsided primary owner. But don't we also get a lopsided primary owner with any other algorithm, just with a different worst-case arrangement?

bhalevy replied:
The problem I was thinking about is distributung thd primary owner work evenly across datacenters and rack, with a small number of tablets.
If there are enough tablets, with the ordinal selection over sorted replicas, by host_id, there shouldn't be any such unfairness.

However, if indeed the tablets are imbalanced and some nodes own considerably more tablets than others, then the bias will be reflected on the primary replicas as well, since on average the number of primary replicas on each node will be the number of tablets on the node divided by the total number of replicas.

bhalevy replied:
If we keep the primary assignment in the map, we can balance the selection by keeping a tally of number of primary replicas per node, similar to the number of replicas per node and shard we use for tablet allocation.
But this will need to get updated on tablet transitions.

bhalevy replied:
I implemented load-balancing approach in v9
by keeping a tally of number of primary replicas per node
so to distributed the primary-replica load evenly across all tablets in a table.
This is not kept anywhere.  It is valid per the immutable tablet_map (that's held alive with the erm/token_metadata_ptr).

avikivity commented:
Missing: some test that proves it works (by looking at logs?)

bhalevy commented:
> Missing: some test that proves it works (by looking at logs?)

There's https://jenkins.scylladb.com/view/master/job/scylla-master/job/tablets/job/gating-dtest-release-with-tablets/75/testReport/repair_additional_test/TestRepairAdditional/FullDtest___full_split004___test_repair_option_pr_multi_dc/
that reliably fails without this fix and reliably passes with it.

avikivity commented:
> > Missing: some test that proves it works (by looking at logs?)
> 
> There's https://jenkins.scylladb.com/view/master/job/scylla-master/job/tablets/job/gating-dtest-release-with-tablets/75/testReport/repair_additional_test/TestRepairAdditional/FullDtest___full_split004___test_repair_option_pr_multi_dc/ that reliably fails without this fix and reliably passes with it.

And the cover letter even mentions it.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9210/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9210/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9210/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1739/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9764/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9210/)
#### Build Details:
- Duration: 3 hr 5 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
In v9 ([6658613](https://github.com/scylladb/scylladb/commit/6658613959d64dde7eaf8fe2da5271b0e83cd571))
- rewrote primary replica selection algorithm to use a load map, tracking the number of primary replicas selected on each node across all tablets in a table, so to distribute the primary replica load evenly across nodes, as much as possible.
- replaces tablet_map::get_primary_replica for storage_service::get_tablet_to_endpoint_map and used for tablet_repair

bhalevy commented:
Converted to draft until I complete testing of the latest version

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9769/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1744/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/)
#### Build Details:
- Duration: 2 hr 40 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
Depends on https://github.com/scylladb/scylladb/pull/18890

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9769/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1744/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9217/)
> 
> #### Build Details:
> * Duration: 2 hr 40 min
> * Builder: spider7.cloudius-systems.com

https://github.com/scylladb/scylla-dtest/issues/3686

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9290/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9290/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1790/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9814/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9290/)
#### Build Details:
- Duration: 3 hr 3 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
@asias @denesb @avikivity please re-review the revised algorithm

denesb commented on locator/tablets.cc:
```diff
@@ -245,9 +245,31 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+                    return node && node->dc_rack().dc == dc;
+                }));
+        if (replicas.empty()) {
+            throw std::runtime_error(format("Could not find any replicas for table={} tablet={} dc={}", table, tid, dc));
```
Why is this an error? Isn't it perfectly legal to configure RF=0 in a certain DC? In that case this method should just return epmty map.

bhalevy replied:
Hmm, good point.
The utility of it is questionable since repairing within a dc with RF=0 is futile, but I agree it doesn't have to cause an error.  I wonder if a warning is appropriate, but it might spam the log since it will be printed for every tablet.

bhalevy replied:
fixed in v11

denesb commented on locator/tablets.cc:
```diff
@@ -245,9 +245,31 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+            throw std::runtime_error(format("Could not find any replicas for table={} tablet={} dc={}", table, tid, dc));
+        }
+        std::ranges::sort(replicas, std::less{}, [&] (const tablet_replica& tr) -> uint64_t {
+            if (auto it = load_map.find(tr.host); it != load_map.end()) {
```
Isn't `load_map` always empty at this point? I only see content being added below.

denesb replied:
Ah, I see this is in the `for_each_tablet()` loop, so it will have content by the second+ iterations.

bhalevy replied:
Maybe we can use a struct to hold the uint64_t value with default initialization to 0 and then
just return `load_map[tr.host]`

bhalevy replied:
used struct default initialization to simplify the sort cmp function v11

denesb commented on locator/tablets.hh:
```diff
@@ -359,8 +360,8 @@ public:
 
-    /// Returns the primary replica for the tablet
-    host_id get_primary_replica(tablet_id id) const;
+    /// Returns the calculated primary replica map for the table
```
This needs more explanation? What is a load map?

denesb replied:
BTW we could have kept the previous interface, with `tablet_map` calculating the load map on-demand and caching it for later uses.
IMO the old interface was more intuitive, but I don't have a strong opinion, this one is fine too.

bhalevy replied:
The problem with caching is when to reuse the cached map and when to evict it.
Assuming that the tablet_map and the topology are immutable, we can store it in the tablet_map.
But we need to make it is cleared when the tablet_map is cloned before we change it.

bhalevy replied:
Also, since for repair we may want a primary replica within a dc, we'll need to cache a primary replica per dc (plus a global one that can just be one of the per-dc primaries).  All in all, it seems overly complicated for a mechanism we currently use only in the control plane.

denesb replied:
I see that caching would make this complicated, so let's keep your current interface.

bhalevy replied:
kept with no caching in v11

bhalevy replied:
also improved the declaration comment (at least tried to make it clearer)

bhalevy commented:
v11 ([7b51670](https://github.com/scylladb/scylladb/commit/7b516704f7f06f503ad3e08d15dc11beaa01c722)):
- simplified load_map implementation using `struct node_load { value = 0; }`
- throw exception on empty replicas only when no dc is specified,
  - otherwise, log a warning if only a subset of the tablets had no replicas in the dc

bhalevy commented:
@denesb we've pretty much reached the end of the day and I don't see any object from @avikivity so I suggest we queue this change once it passes CI since I believe it's stable based on the dtest I've tested locally.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9319/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9319/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1809/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9833/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9319/)
#### Build Details:
- Duration: 6 hr 46 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint please conside merging

asias commented on locator/tablets.cc:
```diff
@@ -245,9 +245,48 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+                return make_ready_future<>();
+            }
+        }
+        std::ranges::sort(replicas, std::less{}, [&] (const tablet_replica& tr) {
```
I think we need std::ranges::stable_sort. The load_map of two nodes could be equal and the order of nodes matters.

bhalevy replied:
since the sort algorithm is symmetrical andcall nodes have the same starting conditions, they should all reach the same result, but maybe we should count on that. Some nodes may run a different version of the stl, for example.
stable_sort is one option, or alternatively the sort function can compare the host_id if the load is equal between any 2 replicas.

asias replied:
I am not sure if the same stl version would give exact the same result on different runs, e.g.,  when the memory situation differs, or when it starts to switch between quick sort or heap sort under the hood. 

Most of the time stable_sort is also n*log(n).  Debugging bug due to unexpected sort would be a nightmare. Using stable_sort makes more sense.

bhalevy replied:
changed to stable_sort in v12

bhalevy commented:
In v12 ([e419e51](https://github.com/scylladb/scylladb/commit/e419e51628921b7994cce4dce9deca961139e6a7)):
- used std::ranges::stable_sort instead of (unstable) sort in calc_primary_replica_map
- rebased

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9357/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9357/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1819/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9842/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9357/)
#### Build Details:
- Duration: 3 hr 6 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint please consider merging

avikivity commented on locator/tablets.cc:
```diff
@@ -245,9 +245,48 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
-host_id tablet_map::get_primary_replica(tablet_id id) const {
-    const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+future<tablet_map::primary_replica_map> tablet_map::calc_primary_replica_map(table_id table, const topology& topo, sstring dc) const {
```
Please use normal words like calculate.

avikivity replied:
Is this function stable? Will it return the same result on all nodes? How much does a single tablet migration affect it? If this is guaranteed, why is it not documented?

bhalevy replied:
The function is stable and will return the same result on all nodes assuming that the topology is the same on all nodes. (and we already know we have to keep it stable throught repair so not to miss repairing a tablet if it migrated under our feet)

That said, the load-balanced selection approach might be more sensitive to single tablet migration since might cause a ripple effect due to difrrent primary selection at this tablet which changes the cumulative load map and therefore affects further selections from that point on (since the per-tablet selections are no longer independent of each other as was the case with the previous algorithm I suggested that was more sensitive to bad tablet ownership imbalance.

avikivity commented on repair/repair.cc:
```diff
@@ -2135,6 +2136,8 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
                 throw std::runtime_error("You need to run primary range repair on all nodes in the cluster.");
             }
+            primary_replicas = co_await tmap.calc_primary_replica_map(tid, erm->get_topology(),
+                    select_primary_ranges_within_dc ? mydc : "");
```
This is very wasteful. This shard (or node?) holds just a few tablets, yet you run a calculation that scans all the tablets for the table.

bhalevy replied:
primary selection is per node, not per shard.
I agree we can skip the selection for the repair use case if there's no replica on the node anyhow.
I'm not sure if that holds true also for the storage_service api case as well.
It might need to know primary ownership on other nodes as well for simulating the vnodes ownership api.

avikivity replied:
What if there's >=1 tablet replica per node?

You're doing O(cluster) work on every node.

bhalevy replied:
> What if there's >=1 tablet replica per node?

Of the same tablet?  That would be a bug.

> You're doing O(cluster) work on every node.

If we persist the primary selection in the table we could calculate it once and reevaluate it on split / merge and maybe migrations.

avikivity replied:
> > What if there's >=1 tablet replica per node?
> 
> Of the same tablet? That would be a bug.

No, I mean every node holds at least one tablet replica. This is expected to be the common case and negates the optimization ("we can skip the selection for the repair use case if there's no replica on the node anyhow").

> 
> > You're doing O(cluster) work on every node.
> 
> If we persist the primary selection in the tablets topology we could calculate it once and reevaluate it on split / merge and maybe migrations.

It should be persisted, and part of the main data structure, not yet another copy.

avikivity replied:
Or calculated with the modulo algorithm, I don't know why you abandoned it.

bhalevy replied:
> > > What if there's >=1 tablet replica per node?
> > 
> > 
> > Of the same tablet? That would be a bug.
> 
> No, I mean every node holds at least one tablet replica. This is expected to be the common case and negates the optimization ("we can skip the selection for the repair use case if there's no replica on the node anyhow").

It would be the common case when RF==|nodes in dc| which might be a common case, but larger installations would have RF < |nodes in dc| and may benefit from the optimization

> > > You're doing O(cluster) work on every node.
> > 
> > 
> > If we persist the primary selection in the tablets topology we could calculate it once and reevaluate it on split / merge and maybe migrations.
> 
> It should be persisted, and part of the main data structure, not yet another copy.

Let's keep in mind that we should select a global primary replica as well as primary per DC.
I wonder if it would be useful in the future if raft groups will be based on the tablet replica where the global and per-dc primary replicas could be used for electing the group global leader and possibly per-dc gateway/coordinator.

> Or calculated with the modulo algorithm, I don't know why you abandoned it.

I abandoned it because of https://github.com/scylladb/scylladb/pull/18784#discussion_r1615262160
I think that your observation is correct that the modulo-based primary selection algorithm is susceptible to bad distribution of replicas.  Therefore I introduced the load_map algorithm that tries to balance the primary-replica load evenly, by providing a counter-bias to the bad distribution.

I'm ok with going back to that for the short term and continue working on persisting the primary ownership as a follow up for the longer term.

kostja replied:
Thinking of LWT, the primary replica has a special role of being the write leader/coordinator, be it paxos or raft. The information is relevant to the clients as well: Scylla's shard-aware drivers direct LWT writes to primaries if they know they are executing a conditional statement.

I'm not sure therefore it's correct to distribute the primaries across based on the load (i.e. number of tablets per node), some nodes may get more heat than others this way. But mainly, the clients should be able to know which tablet replica owner is the primary one, so persisting this information is a good idea.

bhalevy commented:
I've simulated the modulo method (without sorting replicas) over the randomized replica selection (based on 1adf16ce73f135566cd395ace31bb16d41276b12) using 
[tablets-primary-replica-sim.py.txt](https://github.com/user-attachments/files/15526164/tablets-primary-replica-sim.py.txt)
with different simulated topologies and the results seem satisfactory for not starving any node of primary replicas and having low standard deviation (of course small number of tablets and random factors can cause unfair distributions, but when simulating enough (100) random experiments a-la Monte Carlo, the relative standard deviation (divided by the mean) is typically well below 5%.

One crucial factor is the randomized selection of nodes with racks when they have equal tablet replicas load.
Without it, regular selection pattern can "harmonize" with the modulo primary replica selection causing very biased distribution of primary replicas e.g. where primaries are all on odd nodes, while no even node is selected for a primary replica.

Some examples:
```
num_dcs=1 nodes_per_dc=3 racks_per_dc=1 rf_per_dc=3 tablet_count=128 iterations=100
min=4190 mean=4266.67 max=4339 stdev=74.59 conf=0.017

num_dcs=2 nodes_per_dc=6 racks_per_dc=3 rf_per_dc=3 tablet_count=128 iterations=100
min=1024 mean=1066.67 max=1125 stdev=32.28 conf=0.030

num_dcs=1 nodes_per_dc=5 racks_per_dc=1 rf_per_dc=3 tablet_count=128 iterations=100
min=2523 mean=2560.00 max=2595 stdev=32.16 conf=0.013
```

bhalevy commented:
v13 (18df36d):
- back to modulo method (based on good simulation results)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9403/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9403/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9873/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1849/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9403/)
#### Build Details:
- Duration: 2 hr 37 min
- Builder: spider3.cloudius-systems.com
